Add an automatic rebase action

pezholio opened the pull request:
Sometimes, we set up projects to require the branches to be up to date before merging*. With lots of devs on one project at a time and PRs flying everywhere, this can get annoying having to pull down and rebase code.

This adds an action to allow users to add a comment with the `/rebase` command, the code then gets pulled down, rebased and pushed back up.

The only gotcha is that, in order for CI to be kicked off again by the push, we need the repo to have a personal access token set. I've added a check for this, which adds a note to the PR if the repo has a missing access token, with instructions on how to fix this.

\* I do wonder if this is something we might want to set up across all of our projects, but that's probably more one for an RFC

pezholio committed:
Add an automatic rebase action

Sometimes, we set up projects to require the branches to be up to date before
merging. With lots of devs on one project at a time and PRs flying everywhere, this
can get annoying having to pull down and rebase code.

This adds an action to allow users to add a comment with the `/rebase` command,
the code then gets pulled down, rebased and pushed back up.

The only gotcha is that, in order for CI to be kicked off again by the push, we need
the repo to have a personal access token set. I've added a check for this, which
adds a note to the PR if the repo has a missing access token, with instructions on
how to fix this.

Floppy commented:
I thought GitHub had a button for this. Did that go away?

pezholio commented:
It does, but they add a merge commit, which makes the history messy

Floppy commented:
This feels like something we could package up as a reusable action. That might make it easier to reuse on other projects, and be a nice open source thing too.

tahb commented:
Nice automation!

- I wonder if we should use a PAT from the `dxw-rails-user` so we avoid things breaking when people lose access to repos.
- There's a lot of code here. I'd love to test it. Have you been using this somewhere already?

pezholio commented:
@tahb I've been testing it to destruction in [this repo](https://github.com/pezholio/testing-autorebase/). Good call on the token. Do you know if templates can copy across secrets too?

pezholio commented:
@tahb Looking at the settings, administrators can set secrets across the whole org, so that'll help us

pezholio commented:
@tahb I've generated a Personal Access Token for the Rails user and added the token as an organisation-level secret, so all repos have access to it. I've also changed the name of the secret, so it's a bit clearer as to what it exactly is.

pezholio committed:
Trigger rebase action if PR is set to automerge

This builds on the previous commit and adds the `auto_merge_enabled` event
as a trigger to do an automatic rebase, so if a PR is ready to go, but only blocked
by being behind the source branch, the branch will be rebases and then
automatically merged.

erbridge commented:
> I've generated a Personal Access Token for the Rails user and added the token as an organisation-level secret, so all repos have access to it.

That's kind of terrifying to me, given the Rails user has admin access to everything a developer has admin access to. If any repo can access it, we have to be really careful about what we do in every repo (including what GitHub Actions we use).

Can we check for its presence instead, and not run the workflow if it isn't?

pezholio commented:
@erbridge Fair enough. I've removed that now. We already have a [step that checks for the access token and adds a message with instructions to the relevant PR](https://github.com/dxw/rails-template/blob/bdd9b68521c48266f56e6f04fcd6516484a738f6/.github/workflows/rebase.yml#L18) if there is no token, so I think we're good to go

tahb reviewed (approved):
This looks good to me. Let's give it a try!

lozette reviewed (approved):
Happy to give this a try 👍

erbridge commented on .github/workflows/rebase.yml:
```diff
@@ -0,0 +1,72 @@
+jobs:
+  check_token:
+    if:
+      ${{ (github.event.issue.pull_request != '' &&
+      contains(github.event.comment.body, '/rebase')) || github.event_name ==
+      'pull_request_target' }}
```
Pretty sure you don't need the `${{ }}` for `if` values, btw.
